api/src/main/java/org/apache/iceberg/expressions/InclusiveMetricsEvaluator.java

```diff
@@ -269,11 +270,38 @@ public Boolean or(Boolean leftResult, Boolean rightResult) {
 
     @Override
     public <T> Boolean in(BoundReference<T> ref, Set<T> literalSet) {
+      Integer id = ref.fieldId();
+
+      if (containsNullsOnly(id)) {
+        return ROWS_CANNOT_MATCH;
+      }
+
+      final Comparator<T> comparator = ((BoundSetPredicate<T>) expr).comparator();
```

Comment:
This can't use `expr` because that is the expression being evaluated. There is no guarantee it is a bound set predicate.

Instead, I think this should add a `comparator()` method to `BoundReference` because it is passed into all of these methods. That has a `type()` accessor that you can use to get the comparator, like this:

```java
  Comparator<T> comparator() {
    return Comparators.forType(type().asPrimitiveType());
  }
```

Reply:
Done.